pkg/filewatcher/filewatcher.go

```diff
@@ -0,0 +1,179 @@
+// Copyright 2018 Istio Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package filewatcher
+
+import (
+	"fmt"
+	"path/filepath"
+	"sync"
+
+	"github.com/fsnotify/fsnotify"
+	"github.com/hashicorp/go-multierror"
+)
+
+type FileWatcher interface {
+	Add(path string) error
+	Remove(path string) error
+	Close() error
+	Events(path string) chan fsnotify.Event
+	Errors(path string) chan error
+}
+
+type fsNotifyWatcher struct {
+	mu sync.RWMutex
+	// The watcher maintain a map of fsnotify watchers,
+	// keyed by path.
+	watchers map[string]*fsnotify.Watcher
+	// The watcher maintain a map of event channel,
+	// keyed by path.
+	// The watcher watches parent path of given path,
+	// and filter out events of given path, then redirect
+	// to the result channel.
+	// Note that for symlink files, the content in received events
+	// do not have to be related to the file itself.
+	events map[string]chan fsnotify.Event
+}
+
+// NewWatcher return with a FileWatcher instance that implemented with fsnotify.
+func NewWatcher() FileWatcher {
+	w := &fsNotifyWatcher{
+		watchers: map[string]*fsnotify.Watcher{},
+		events:   map[string]chan fsnotify.Event{},
+	}
+
+	return w
+}
+
+// Add is implementation of Add interface of FileWatcher.
+func (w *fsNotifyWatcher) Add(path string) error {
+	w.mu.Lock()
+	defer w.mu.Unlock()
+
+	if _, exist := w.watchers[path]; exist {
+		return fmt.Errorf("path %s already added", path)
+	}
+
+	watcher, err := fsnotify.NewWatcher()
+	if err != nil {
+		return err
+	}
+
+	events := make(chan fsnotify.Event, 1)
+
+	cleanedPath := filepath.Clean(path)
+	parentPath, _ := filepath.Split(cleanedPath)
+	// Ignore the error here as the file can be an "usual" file.
+	realPath, _ := filepath.EvalSymlinks(path)
+
+	watcher.Add(parentPath)
+
+	go func() {
+		for {
+			select {
+			case event, ok := <-watcher.Events:
+				if !ok { // 'Events' channel is closed
+					return
+				}
+				// Ignore the error here as the file can be an "usual" file.
+				currentRealPath, _ := filepath.EvalSymlinks(path)
+
+				// Filter out events that related to the path to watch:
+				// for usual files, event name should be equal to its path;
+				// For k8s configmap/secret file, real path of the symlink
+				// would change on update.
+				if filepath.Clean(event.Name) == cleanedPath ||
```

Comment:
Wouldn't this cause change events to be missed when the sym-link target doesn't change, but the targeted file itself changes?

Reply:
That's indeed the issue.
I thought we would mainly focus on the k8s congfigmap/secret scenario (as this will not happen to congfigmap/secret). 

But as a general tool, it is indeed not comprehensive enough. Then we can switch to the `checksum` approach @ayj suggested. However, I am kind of concerned about its performance.
But considering that config might not be updated frequently, it should be acceptable.